### PR TITLE
LB-305: Improve parameter checking for timestamps

### PR DESCRIPTION
**Pull request by paramsingh**
# Summary

* This is a…
    * [ ] Bug fix
    * [x] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

* JIRA ticket (_optional_): [LB-305](https://tickets.metabrainz.org/browse/LB-305) 

# Solution

If the timestamp is greater than the time at our server, the listen should be invalid.

Also, added tests.

### CODE REVIEW

**Comment by paramsingh** [comment]
Based this off of production because it is needed there too. We could do a merge into master after this gets merged?

File: listenbrainz/webserver/views/api_tools.py [hunk 1]
```diff
@@ -104,6 +105,9 @@ def validate_listen(listen, listen_type):
             log_raise_400("JSON document may only contain listened_at and "
                            "track_metadata top level keys", listen)
 
+        if listen['listened_at'] > int(time.time()):
```
Comment (alastair) [inline listenbrainz/webserver/views/api_tools.py]: perhaps it's a good idea to consider a bit of clock skew. I'd consider allowing anything from 10 minutes to 1h in advance

File: listenbrainz/tests/integration/test_api.py [hunk 2]
```diff
@@ -291,3 +291,14 @@ def test_multiple_artist_names(self):
         self.assert400(response)
         self.assertEqual(response.json['code'], 400)
         self.assertEqual('artist_name must be a single string.', response.json['error'])
+
+    def test_too_high_timestamps(self):
```
Comment (mayhem) [inline listenbrainz/tests/integration/test_api.py]: https://www.youtube.com/watch?v=kcsNbQRU5TI

File: listenbrainz/webserver/views/api_compat_deprecated.py [hunk 3]
```diff
@@ -156,8 +156,14 @@ def _to_native_api(data, append_key):
...
-            listen['listened_at'] = data['i{}'.format(append_key)]
-        except KeyError:
+            listen['listened_at'] = int(data['i{}'.format(append_key)])
+        except (KeyError, ValueError):
+            return None
+
+        # if timestamp is too high, this is an invalid listen
+        # in order to make up for possible clock skew, we allow
+        # timestamps to be one hour ahead of server time
+        if listen['listened_at'] > int(time()) + 86400:
```
Comment (mayhem) [inline listenbrainz/webserver/views/api_compat_deprecated.py]: Now we have two timestamp upper limit checkers, with different values. Why is this one 24 hours? I think we should have one function for this.

File: listenbrainz/webserver/views/api_tools.py [hunk 4]
```diff
@@ -104,6 +105,12 @@ def validate_listen(listen, listen_type):
             log_raise_400("JSON document may only contain listened_at and "
                            "track_metadata top level keys", listen)
 
+        # if timestamp is too high, raise BadRequest
+        # in order to make up for possible clock skew, we allow
+        # timestamps to be one hour ahead of server time
+        if listen['listened_at'] > int(time.time()) + 60 * 60:
```
Comment (mayhem) [inline listenbrainz/webserver/views/api_tools.py]: Here is the other instance, at 1 hour.

File: listenbrainz/webserver/views/api_compat_deprecated.py (hunk 3 above)
Comment (paramsingh) [reply listenbrainz/webserver/views/api_compat_deprecated.py]: whoops.

File: listenbrainz/webserver/views/api_tools.py [hunk 5]
```diff
@@ -104,6 +106,12 @@ def validate_listen(listen, listen_type):
             log_raise_400("JSON document may only contain listened_at and "
                            "track_metadata top level keys", listen)
 
+        # if timestamp is too high, raise BadRequest
+        # in order to make up for possible clock skew, we allow
+        # timestamps to be one hour ahead of server time
+        if listen['listened_at'] > int(time.time()) + API_LISTENED_AT_ALLOWED_SKEW:
```
Comment (mayhem) [inline listenbrainz/webserver/views/api_tools.py]: Sorry, I should've seen this earlier, but this allows for only a clock skew for the users clock to be FAST, not SLOW. I think we should test for 30 minutes fast and or 30 minutes slow.

Comment (paramsingh) [reply listenbrainz/webserver/views/api_tools.py]: If I do  `current_time - API_LISTENED_AT_ALLOWED_SKEW <= ts <= currentt_time + API_LISTENED_AT_ALLOWED_SKEW`, it wouldn't work for old listens leading to failing imports.

Comment (mayhem) [reply listenbrainz/webserver/views/api_tools.py]: Sorry, can you please elaborate? I don't understand what you're saying.

**Pull request merged**